### PR TITLE
DRILL-4946: redirect System.err so users under embedded mode won't se…

### PR DESCRIPTION
**Pull request by chunhui-shi**
…e error stacks printed due to errors occurred in scalar replacement.

### CODE REVIEW

File: exec/java-exec/src/main/java/org/apache/drill/exec/util/StdErrCapturer.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+
+import java.io.ByteArrayOutputStream;
+import java.io.OutputStream;
+import java.io.PrintStream;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * The safe way of this class is to redirect System.err to nullOutputStream
+ * If
```
Comment (chunhui-shi) [inline exec/java-exec/src/main/java/org/apache/drill/exec/util/StdErrCapturer.java]: The comment is now finished.

File: exec/java-exec/src/main/java/org/apache/drill/exec/util/StdErrCapturer.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+import java.io.PrintStream;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * The safe way to use this class is to redirect System.err to nullOutputStream.
+ * If callers want to get the redirected output -ito print or log elsewhere - as used in unit tests and they
+ * need the output to be neat, then there should be advanced ways to preserve the output to avoid concurrent 
+ * inputs messing up. But this class is not designed for that.
+ */
+public class StdErrCapturer {
```
Comment (paul-rogers) [inline exec/java-exec/src/main/java/org/apache/drill/exec/util/StdErrCapturer.java]: If an exception occurs, this class can leave the System.stderr redirected, causing mysterious errors. Suggestion: make this AutoClosable: 

try ( StdErrCapturer cap = new StdErrCapturer( ) ) {
  // Do your stuff here
}

The above will ensure that stderr is restored whether the enclosed code succeeds or fails.

File: exec/java-exec/src/main/java/org/apache/drill/exec/util/StdErrCapturer.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  static private PrintStream savedStdErr;
+  static private OutputStream buffer;
+  static private PrintStream capturingStream;
+  static Boolean redirected = false;
+
+  public static void redirect() {
```
Comment (paul-rogers) [inline exec/java-exec/src/main/java/org/apache/drill/exec/util/StdErrCapturer.java]: Do in constructor (so it can be reversed in close.)

**Comment by ilooner** [comment]
@chunhui-shi @paul-rogers Since Drill has moved to Java 8, which is supposed to do scalar replacement for us, would it be appropriate to commit this change to turn off scalar replacement? Or is this change no longer applicable?

**Comment by chunhui-shi** [comment]
I suggest to set scalar replacement flag to off by default. I think we have compared the performance between on/off before.
------------------------------------------------------------------
Sender:Timothy Farkas <notifications@github.com>
Sent at:2018 Jun 21 (Thu) 10:45
To:apache/drill <drill@noreply.github.com>
Cc:Chunhui Shi <shi.chunhui@aliyun.com>; Mention <mention@noreply.github.com>
Subject:Re: [apache/drill] DRILL-4946: redirect System.err so users under embedded mode won't se… (#619)

@chunhui-shi @paul-rogers Since Drill has moved to Java 8, which is supposed to do scalar replacement for us, would it be appropriate to commit this change to turn off scalar replacement? Or is this change no longer applicable?

**Comment by paul-rogers** [comment]
@ilooner, yes, the change here seems fine. We're just passing an option to Sqlline (which passes it to Drill). This is a cleaner way to avoid the original issue. So,
+1

**Comment by vvysotskyi** [comment]
I don't think it is the correct decision to disable scalar replacement for embedded mode only just because of printing stack trace to stderr. 

Also, I think making more differences in work between embedded and distributed mode is not a good idea. 

If we want to disable scalar replacement, we should do that independently of running mode or used client.

**Comment by ilooner** [comment]
Fair point @vvysotskyi . @chunhui-shi Would you have time to disable scalar replacement for all modes?

**Comment by vvysotskyi** [comment]
Closing this PR, since most of the problems which caused errors during scalar replacement were resolved.

**Pull request closed without merging**